pytential/symbolic/matrix.py

```diff
@@ -557,40 +570,43 @@ def map_int_g(self, expr):
-        if is_zero(rec_density):
-            return 0
+        result = 0
+        for kernel, density in zip(expr.source_kernels, expr.densities):
+            rec_density = self._blk_mapper.rec(density)
+            if is_zero(rec_density):
+                continue
 
-        if not np.isscalar(rec_density):
-            raise NotImplementedError
+            if not np.isscalar(rec_density):
+                raise NotImplementedError
 
-        # NOTE: copied from pytential.symbolic.primitives.IntG
-        # NOTE: P2P evaluation only uses the inner kernel, so it should not
-        # get other kernel_args, e.g. normal vectors in a double layer
-        kernel = expr.kernel.get_base_kernel()
-        kernel_args = kernel.get_args() + kernel.get_source_args()
-        kernel_args = {arg.loopy_arg.name for arg in kernel_args}
+            # NOTE: copied from pytential.symbolic.primitives.IntG
+            kernel_args = kernel.get_args() + kernel.get_source_args()
+            kernel_args = {arg.loopy_arg.name for arg in kernel_args}
 
-        actx = self.array_context
-        kernel_args = _get_layer_potential_args(self._mat_mapper,
-                expr, include_args=kernel_args)
-        if self.exclude_self:
-            kernel_args["target_to_source"] = actx.from_numpy(
-                    np.arange(0, target_discr.ndofs, dtype=np.int)
-                    )
-
-        from sumpy.p2p import P2PMatrixBlockGenerator
-        mat_gen = P2PMatrixBlockGenerator(actx.context, (kernel,),
-                exclude_self=self.exclude_self)
-
-        from meshmode.dof_array import flatten, thaw
-        _, (mat,) = mat_gen(actx.queue,
-                targets=flatten(thaw(actx, target_discr.nodes())),
-                sources=flatten(thaw(actx, source_discr.nodes())),
-                index_set=self.index_set,
-                **kernel_args)
-
-        return rec_density * actx.to_numpy(mat)
+            actx = self.array_context
+            kernel_args = _get_layer_potential_args(self._mat_mapper,
+                    expr, include_args=kernel_args)
+            if self.exclude_self:
+                kernel_args["target_to_source"] = actx.from_numpy(
+                        np.arange(0, target_discr.ndofs, dtype=np.int)
+                        )
+
+            from sumpy.p2p import P2PMatrixBlockGenerator
+            mat_gen = P2PMatrixBlockGenerator(actx.context,
+                    source_kernels=(kernel,),
+                    target_kernels=(expr.target_kernel,),
+                    exclude_self=self.exclude_self)
+
+            from meshmode.dof_array import flatten, thaw
+            _, (mat,) = mat_gen(actx.queue,
+                    targets=flatten(thaw(actx, target_discr.nodes())),
+                    sources=flatten(thaw(actx, source_discr.nodes())),
+                    index_set=self.index_set,
+                    **kernel_args)
+
+            result += rec_density * actx.to_numpy(mat)
+
+        return result
```

Comment:
@alexfikl Could you take a look at this? See https://github.com/inducer/sumpy/pull/23/files for context?

Reply:
Looks good to me!

What's a source/target kernel? (Couldn't find any sufficiently dumbed-down docs :(

Reply:
Thanks for having a look. Here's an explanation that I just added, https://github.com/inducer/pytential/blob/392afb3db03cf72991d5acd01d28f89e7e5cbb2c/pytential/symbolic/primitives.py#L1424
Let me if that is sufficient. If not, I can improve the docstring.

Reply:
That clears it up for me. Thanks!